### ISSUE TITLE
avoiding allocations, copies

### ISSUE THREAD

**Issue by will** [will/native_pg_driver#1]
Hey @asterite, if you have some time to spare, I was hoping you could look over the start of this to make sure I don’t go down the wrong path. I'd like to avoid allocations, gc pressure, etc as much as possible in the low parts of the driver, and you're pretty good at that :)

Starting from the bottom
https://github.com/will/native_pg_driver/blob/766be84a471b25007d8dfb60ea9beff320ac6dc7/native.cr#L212 issues the query and takes a block to iterate over each value

Then in that method https://github.com/will/native_pg_driver/blob/766be84a471b25007d8dfb60ea9beff320ac6dc7/native.cr#L62 it sends the query and gets a frame, 

then goes into https://github.com/will/native_pg_driver/blob/766be84a471b25007d8dfb60ea9beff320ac6dc7/native.cr#L22 which yields each value one at a time.

I think because it's all yields everything should be inlined, so there's no copying going on but want to double check with you.

Also, trying to think ahead to plugging this into the union-collapsing `exec({String,Int}, "select..")` interface I have, I think something on top of this single-value-yeilding thing could read in an entire row at a time, and do the casting. 

Thanks in advance for your considerations and time.

**Comment by asterite** [comment]
Yes, the only allocation I see is `Slice.new`, but I guess that's reasonable.

By the way, I think you should really try to make it work with [crystal-db](https://github.com/bcardiff/crystal-db), so you'll only have to implement the bare minimum of the protocol and you won't have to worry about reading tuples (the general API will take care of that). Some weeks ago I saw @bcardiff, @waj and @spalladino discussing how to read "non-standard" types from a DB and they'll add that functionality (through a `read_object(T)` method, I think), so that will be covered too, in case that was your worry.

**Comment by will** [comment]
I'll take a look at crystal-db but I really worry about agnostic db things because I haven't seen one that limits everyone down to the least-common-denominator

**Comment by asterite** [comment]
@will We can try to be the first ones :-)

Or maybe you can conform to the API but still use the driver classes without necessarily sitting behind the generic API. This will definitely need a few iterations until we get it right.

**Comment by will** [comment]
Yeah it's worth a shot for sure
